File: platform/app/schemas/task.py
from __future__ import annotations

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, field_validator


class TaskCreateRequest(BaseModel):
    jira_id: Optional[str] = None
    title: str
    description: Optional[str] = None
    template_id: Optional[str] = None
    project_id: Optional[str] = None
    target_branch: Optional[str] = None
    yunxiao_task_id: Optional[str] = None


class TaskStageResponse(BaseModel):
    id: str
    task_id: str
    stage_name: str
    agent_role: str
    status: str
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    duration_seconds: Optional[float] = None
    tokens_used: int = 0
    turns_used: int = 0
    self_fix_count: int = 0
    output_summary: Optional[str] = None
    error_message: Optional[str] = None
    # Phase 1.1: Structured output
    output_structured: Optional[dict] = None
    # Phase 1.2: Failure classification
    failure_category: Optional[str] = None
    # Phase 2.2: Self-assessment confidence
    self_assessment_score: Optional[float] = None
    # Phase 2.5: Per-stage retry count
    retry_count: int = 0

    @field_validator("tokens_used", "turns_used", "self_fix_count", "retry_count", mode="before")
    @classmethod
    def _none_to_zero(cls, v: object) -> int:
        return v if v is not None else 0

    model_config = {"from_attributes": True}


class TaskDetailResponse(BaseModel):
    id: str
    jira_id: Optional[str] = None
    title: str
    description: Optional[str] = None
    status: str
    total_tokens: int = 0
    total_cost_rmb: float = 0.0
    created_at: datetime
    completed_at: Optional[datetime] = None
    branch_name: Optional[str] = None
    pr_url: Optional[str] = None
    stages: List[TaskStageResponse] = []
    template_id: Optional[str] = None
    project_id: Optional[str] = None
    template_name: Optional[str] = None
    project_name: Optional[str] = None
    target_branch: Optional[str] = None
    yunxiao_task_id: Optional[str] = None

    @field_validator("total_tokens", mode="before")
    @classmethod
    def _tokens_none(cls, v: object) -> int:
        return v if v is not None else 0

    @field_validator("total_cost_rmb", mode="before")
    @classmethod
    def _cost_none(cls, v: object) -> float:
        return v if v is not None else 0.0

    model_config = {"from_attributes": True}


class TaskListResponse(BaseModel):
    items: List[TaskDetailResponse]
    total: int
    page: int
    page_size: int


# --- PRD Decompose ---

class TaskDecomposeRequest(BaseModel):
    prd_text: str
    project_id: Optional[str] = None
    template_id: Optional[str] = None


class DecomposedTask(BaseModel):
    title: str
    description: str
    priority: str = "medium"


class TaskDecomposeResponse(BaseModel):
    tasks: List[DecomposedTask]
    summary: str
    tokens_used: int = 0


# --- Batch Create ---

class BatchTaskItem(BaseModel):
    title: str
    description: Optional[str] = None
    template_id: Optional[str] = None
    project_id: Optional[str] = None
    target_branch: Optional[str] = None
    yunxiao_task_id: Optional[str] = None


class TaskBatchCreateRequest(BaseModel):
    tasks: List[BatchTaskItem]


class TaskBatchCreateResponse(BaseModel):
    created: int
    tasks: List[TaskDetailResponse]


class TaskRetryFromStageRequest(BaseModel):
    """Request payload for retrying a task from a specific failed stage."""

    stage_id: str


class TaskBatchRetryRequest(BaseModel):
    """Request payload for retrying multiple tasks."""

    task_ids: List[str]


class TaskBatchRetryItem(BaseModel):
    """Per-task retry result for batch retry operations."""

    task_id: str
    success: bool
    reason: Optional[str] = None
    task: Optional[TaskDetailResponse] = None


class TaskBatchRetryResponse(BaseModel):
    """Aggregated response for batch retry operations."""

    total: int
    succeeded: int
    failed: int
    items: List[TaskBatchRetryItem]

File: web/src/types/task.ts
export interface StageOutputStructured {
  summary: string;
  status: 'pass' | 'fail' | 'partial';
  confidence: number;
  artifacts: string[];
  metadata: Record<string, unknown>;
  // Stage-specific fields (varies by stage)
  [key: string]: unknown;
}

export interface TaskStage {
  id: string;
  task_id: string;
  stage_name: string;
  agent_role: string;
  status: 'pending' | 'running' | 'completed' | 'failed' | 'skipped';
  started_at: string | null;
  completed_at: string | null;
  duration_seconds: number | null;
  tokens_used: number;
  turns_used: number;
  output_summary: string | null;
  error_message: string | null;
  // Phase 1.1: Structured output
  output_structured: StageOutputStructured | null;
  // Phase 1.2: Failure classification
  failure_category: string | null;
  // Phase 2.2: Self-assessment confidence
  self_assessment_score: number | null;
  // Phase 2.5: Per-stage retry count
  retry_count: number;
}

export interface Task {
  id: string;
  jira_id: string | null;
  title: string;
  description: string;
  status: 'pending' | 'running' | 'completed' | 'failed' | 'cancelled' | 'planning';
  created_at: string;
  completed_at: string | null;
  stages: TaskStage[];
  total_tokens: number;
  total_cost_rmb: number;
  template_id: string | null;
  project_id: string | null;
  template_name: string | null;
  project_name: string | null;
  target_branch: string | null;
  yunxiao_task_id: string | null;
  error_reason: string | null;
}

export interface YunxiaoTask {
  id: string;
  title: string;
  status: string;
}

export interface TaskListResponse {
  tasks: Task[];
  total: number;
  page: number;
  page_size: number;
}

export interface TaskCreateRequest {
  title: string;
  description: string;
  template_id?: string;
  project_id?: string;
  target_branch?: string;
  yunxiao_task_id?: string;
}

// --- PRD Decompose ---

export interface TaskDecomposeRequest {
  prd_text: string;
  project_id?: string;
  template_id?: string;
}

export interface DecomposedTask {
  title: string;
  description: string;
  priority: string;
}

export interface TaskDecomposeResponse {
  tasks: DecomposedTask[];
  summary: string;
  tokens_used: number;
}

// --- Batch Create ---

export interface BatchTaskItem {
  title: string;
  description?: string;
  template_id?: string;
  project_id?: string;
  target_branch?: string;
  yunxiao_task_id?: string;
}

export interface TaskBatchCreateRequest {
  tasks: BatchTaskItem[];
}

export interface TaskBatchCreateResponse {
  created: number;
  tasks: Task[];
}

/**
 * Request payload for retrying a task from a specific failed stage.
 */
export interface TaskRetryFromStageRequest {
  /** Target failed stage identifier belonging to the task. */
  stage_id: string;
}

/**
 * Request payload for batch retrying failed tasks.
 */
export interface TaskBatchRetryRequest {
  /** Task identifiers to retry in batch. */
  task_ids: string[];
}

/**
 * Per-task execution result returned by batch retry API.
 */
export interface TaskBatchRetryItem {
  /** Task identifier for this retry item. */
  task_id: string;
  /** Whether retry was accepted for this task. */
  success: boolean;
  /** Failure reason when retry is rejected. */
  reason?: string | null;
  /** Latest task snapshot when backend includes it. */
  task?: Task | null;
}

/**
 * Batch retry summary returned by backend.
 */
export interface TaskBatchRetryResponse {
  /** Number of tasks requested for retry. */
  total: number;
  /** Number of tasks successfully retried. */
  succeeded: number;
  /** Number of tasks failed to retry. */
  failed: number;
  /** Per-task retry result details. */
  items: TaskBatchRetryItem[];
}
